Replace PRNG with additional RO function (#44)
chris-wood/draft-irtf-cfrg-voprf: draft-sullivan-cfrg-oprf.md

## Modified protocol execution

The VOPRF protocol from Section {{protocol}} changes to allow specifying

> multiple blinded PRF inputs [ Mi ] for i in 1...n. Then P computes the array [Zi

Reviewer: ```suggestion
multiple blinded PRF inputs [ Mi ] for i in 1...n. Then P computes the array [Zi]
```

> ] and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The

Reviewer: ```suggestion
and replaces DLEQ_Generate with Batched_DLEQ_Generate over these arrays. The
```

same applies to the algorithm VOPRF_Sign. The same applies for replacing
DLEQ_Verify with Batched_DLEQ_Verify when V verifies the response from P and
during the algorithm VOPRF_Verify.

> ## Random oracle instantions for proofs

Reviewer: ```suggestion
## Random oracle instantiations for proofs
```

We can instantiate the random oracle function H_4 using the same hash function
that is used for H_1,H_2,H_3. For H_5, we can also use a similar instantiation,

> or we can use a stream-based output generator. For example, for groups with an

Reviewer: ```suggestion
or we can use a variable-length output generator. For example, for groups with an
```

order of 256-bit, valid instantiations include functions such as SHAKE-256

> {{SHAKE}} or HKDF_Expand-SHA256 {{RFC5869}}.

Reviewer: Is one preferred over the other? Why? (We should probably include some rationale text here.)

Author: I don't think either is preferred, they're just examples of functions that you might want to use. We give ciphersuites below that use HKDF_Expand, is that enough?

In addition if a function with larger output than the order of the base field is
used, we note that the outputs of H_5 (d1,...,dn) must be smaller than this
order. If any di that is sampled is larger than then order, then we should
resample until a di' is sampled that is valid.

> In these cases, the iterating integer i is increased to i' until such di' is

Reviewer: ```suggestion
In these cases, the iterating integer i is increased monotonically to i' until such di' is
```

> In these cases, the iterating integer i is increased to i' until such di' is

Reviewer: In general, this is underspecified. We should leave a TODO here at least to write this out.

Author: Good point, I will add this.

- H_2: SHA256
- H_3: SHA256
- H_4: SHA256

> - H_5: HKDF_Expand-SHA256

Reviewer: It's typically written as `HKDF-Expand`, not `HKDF_Expand`. Can we use that instead?

Author: Sure

> Z_1: 04647e1ab7946b10c1c1c92dd333e2fc9e93e85fdef5939bf2f376ae859248513e0cd91115\

Reviewer: Are these test vectors correct with HKDF? We should probably remove them, temporarily, if not.

Author: Yeah they are correct with HKDF (rather than Shake)
